docs/reference-manual/native-image/ReplayBundles.md

```diff
@@ -0,0 +1,207 @@
+
+2. Another angle is provided from the development phase. If image building fails or a malfunctioning image is
+created (i.e. the same application runs fine when executed via JVM) we would like to get bug reports that allow us to
+reproduce the problem locally without hours of replicating their setup. We would want some way to bundle up what the
+user built (or tried to build) into a nice package that allows us to instantly reproduce the problem on our side.
+
+3. Debugging an image created long time ago is also sometimes needed. It would be great if there is a single bundle that
+contains everything needed to perform this task.
+
+## Replay Bundles 
+
+A set of options should be added to the `native-image` command that allows to create so-called "replay bundles" that can
+be used to help with problems described above. There shall be
+
+```shell
+native-image --replay-create ...other native-image arguments...
+```
+
+This will instruct native-image to create a replay bundle alongside the image.
+
+For example, after the running:
+
+```shell
+native-image --replay-create -Dlaunchermode=2 -EBUILD_ENVVAR=env42 \
+ -p somewhere/on/my/drive/app.launcher.jar:/mnt/nfs/server0/mycorp.base.jar \
+ -cp $HOME/ourclasses:somewhere/logging.jar:/tmp/other.jar:aux.jar \
+ -m app.launcher/paw.AppLauncher alaunch
+```
+the user sees the following build results:
+```shell
+~/foo$ ls
+alaunch alaunch.nirb.jar somewhere aux.jar
+```
+As we can see, in addition to the image a `<imagename>.nirb.jar`-file was created. This is the native image replay
+bundle for the image that got built. At any time later, if the same version of GraalVM is used, the image can be rebuilt
+with:
+
+```shell
+native-image --replay .../path/to/alaunch.nirb.jar
+```
+
+this will rebuild the `alaunch` image with the same image arguments, environment variables, system properties
+settings, classpath and module-path options as in the initial build.
+
+To support the use-case of image-building-as-a-service, there should also be a way to create a bundle without
+performing the initial build locally. This allows users to offload image building to a cloud-service specialized in
+image building and retaining of replay bundles. The command line for that should be:     
+
+```shell
+native-image --replay-prepare ...other native-image arguments...
+```
+
+## Replay Bundles File Format
+
+A `<imagename>.nirb.jar` file is a regular jar-file that contains all information needed to replay a previous build.
+For example, the `alaunch.nirb.jar` replay bundle has the following inner structure:
+
+```
+alaunch.nirb.jar
+├── build
+│   └── report <- Contains information about the build proccess.
+│       │         In case of replay these will be compared against. 
+│       ├── analysis_results.json
+│       ├── build_artifacts.json
+│       ├── build.log
+│       ├── build_output.json
+│       ├── jni_access_details.json
+│       └── reflection_details.json
+├── input
```

Comment:
- I think JSON config files should be included here (JNI/reflection/resources) etc. Those can be passed via command line options
- Does it make sense to capture compiler/linker versions? There are also dynamic libs, e.g. an AWT program can link `libX11.so` etc, but this is probably too much. But might make sense to capture the linker command that gets executed at the final step

Reply:
> I think JSON config files should be included here (JNI/reflection/resources) etc. Those can be passed via command line options

@petermz good point! All the config files that are passed to native image that were not embedded into on of the input jars need to be put in here as well. We should have an extra subdirectory for such files/directories e.g. `input/auxiliary` would make sense.

> Does it make sense to capture compiler/linker versions?

For non-containerized builds it does. We mostly already have that info in the builder (see class `com.oracle.svm.hosted.c.codegen.CCompilerInvoker`)  we just need to make sure its a mandatory part of `build_output.json`.